sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraIO.java

```diff
@@ -0,0 +1,502 @@
+ *
+ * <h3>Reading from Apache Cassandra</h3>
+ *
+ * <p>{@code CassandraIO} provides a source to read and returns a bounded collection of
+ * entities as {@code PCollection<Entity>}. An entity is built by Cassandra mapper based on a
+ * POJO containing annotations.
+ *
+ * <p>To configure a Cassandra source, you have to provide the hosts, port, and keyspace of the
+ * Cassandra instance, wrapped as a {@link ConnectionConfiguration}. The following example
+ * illustrate various options for configuring the IO:
+ *
+ * <pre>{@code
+ * pipeline.apply(CassandraIO.<Person>read()
+ *     .withConnectionConfiguration(CassandraIO.ConnectionConfiguration.create(
+ *        Arrays.asList("host1", "host2"),
+ *        9042,
+ *        "beam"))
+ *     .withTable("Person")
+ *     .withEntity(Person.class)
+ *     .withCoder(SerializableCoder.of(Person.class))
+ *     // above options are the minimum set, returns PCollection<Person>
+ *
+ * }</pre>
+ */
+public class CassandraIO {
+
+  private static final Logger LOG = LoggerFactory.getLogger(CassandraIO.class);
+
+  private CassandraIO() {}
+
+  public static <T> Read<T> read() {
+    return new AutoValue_CassandraIO_Read.Builder<T>().build();
+  }
+
+  /** POJO describing a connection to Apache Cassandra database connection configuration. */
+  @AutoValue
+  public abstract static class ConnectionConfiguration implements Serializable {
+    public static ConnectionConfiguration create(List<String> hosts, String keyspace, int port) {
+      checkArgument(
+          hosts != null,
+          "ConnectionConfiguration.create(hosts, keyspace, port) " + "called with null hosts");
+      checkArgument(
+          hosts.size() >= 1,
+          "ConnectionConfiguration.create(hosts, keyspace, port) "
+              + "called with an empty hosts list");
+      checkArgument(
+          keyspace != null,
+          "ConnectionConfiguration.create(hosts, keyspace, port) " + "called with null keyspace");
+      checkArgument(
+          port > 0,
+          "ConnectionConfiguration.create(hosts, keyspace, port) called "
+              + "with invalid port number");
+      return new AutoValue_CassandraIO_ConnectionConfiguration.Builder()
+          .setHosts(hosts)
+          .setKeyspace(keyspace)
+          .setPort(port)
+          .build();
+    }
+
+    abstract List<String> getHosts();
+
+    abstract String getKeyspace();
+
+    abstract int getPort();
+
+    abstract Builder builder();
+
+    public void populateDisplayData(DisplayData.Builder builder) {
+      builder.addIfNotNull(DisplayData.item("hosts", getHosts().toString()));
+      builder.addIfNotNull(DisplayData.item("keyspace", getKeyspace()));
+      builder.add(DisplayData.item("port", getPort()));
+    }
+
+    Cluster getCluster() {
+      LOG.debug("Connecting to Cassandra cluster");
+      return Cluster.builder()
+          .addContactPoints(getHosts().toArray(new String[0]))
+          .withPort(getPort())
```

Comment:
Yup - doesn't use keyspace. Keyspace definitely doesn't belong in this class, then.